### PR TITLE
Qualcomm AI Engine Direct - Refactor llama runner for dynamic IO dtypes

### PR DESCRIPTION
**Pull request by chenweng-quic**
### Summary
To enable GPU backend support in the Llama runner, refactoring is required because the dtypes of kv_cache, attention_mask, and logits are currently hardcoded, preventing floating‑point models from running.
This PR focuses on removing the hardcode dtype for them.

#### Key changes
- Remove template parameter <typename T> from KVManager, LhdTokenGenerator,
  MultimodalPromptProcessor, and related runner classes
- Detect kv_cache and attention_mask dtypes dynamically from MethodMeta at
  construction time instead of compile-time bitwidth detection
- Switch to std::byte* pointer arithmetic with getDtypeSize() for all buffer
  offsets; add fill_mask() helper for multi-dtype attention mask filling
- Update spec_prop pass for custom llama op for sharding case greater than 1

### Test plan
```
python backends/qualcomm/tests/test_qnn_delegate.py -k TestExampleLLMScript.test_llama_stories_110m --model SM8650 --build_folder /local/mnt/workspace/chenweng/executorch/executorch/build-android  --device acfa9311 --executorch_root . --artifact_dir ./stories_110m_pte_size --llama_artifacts . --use_fp16
```
image

cc @cccclai @cbilgin @abhinaykukkadapu

### CODE REVIEW

**Comment by chenweng-quic** [comment]
Hi @abhinaykukkadapu,
This PR touch the spec_prop_pass since the spec written in build_quant_io is restore by spec_prop_pass when num_sharding greater than 1. 
e.g.,
```
after build_quant_io
input -> call_delegate1 (spec=uint16) -> llama.fallback (spec=uint16) -> call_delegate2 (spec=uint16) -> output
after spec_prop_pass 
input -> call_delegate1 (spec=uint16) -> llama.fallback (spec=float) -> call_delegate2 (spec=uint16) -> output
```

It will cause following error in op_fallback.cpp
image

To fix the dtype issue, I add logic in spec_prop_pass to avoid restoring.
I am not sure if this is correct fix, could you share your opinion?

**Comment by chenweng-quic** [comment]
Hi @abhinaykukkadapu,
Could you tag claude to review this PR?
Thanks

**Comment by abhinaykukkadapu** [comment]
@claude review this PR

**Comment by chenweng-quic** [comment]
Hi @psiddh,
Could you take a look of this PR?
Thanks.

**Comment by psiddh** [comment]
@claude Re-review the PR thoroughly

**Comment by psiddh** [comment]
Looks like this PR has some conflicts., Please take a look at CI failures , one of them seem related to this PR. 
Take a look at the new issues raised by Claude. I will review this further tomorrow

**Comment by chenweng-quic** [comment]
Hi @psiddh,
I fixed the CI failures, could you review again?
Thanks

File: examples/qualcomm/oss_scripts/llama/runner/utils.h [hunk 1]
```diff
@@ -20,3 +26,38 @@ struct TensorStruct {
   // data size in bytes
   size_t size;
 };
+
+inline size_t getDtypeSize(executorch::aten::ScalarType dtype) {
+  switch (dtype) {
```
Comment (psiddh) [inline examples/qualcomm/oss_scripts/llama/runner/utils.h:31]: Should we also add 
case exec_aten::ScalarType::Half:

File: examples/qualcomm/oss_scripts/llama/runner/decoder_runner.h [hunk 2]
```diff
@@ -56,19 +57,36 @@ class DecoderRunner {
...
+        case executorch::aten::ScalarType::UInt16:
+          logits_f[i] = reinterpret_cast<uint16_t*>(logits_last)[i];
+          break;
+        case executorch::aten::ScalarType::Byte:
+          logits_f[i] = reinterpret_cast<uint8_t*>(logits_last)[i];
+          break;
+        case executorch::aten::ScalarType::Float:
+          logits_f[i] = reinterpret_cast<float*>(logits_last)[i];
+          break;
+        default:
```
Comment (psiddh) [inline examples/qualcomm/oss_scripts/llama/runner/decoder_runner.h:84]: Should we also add something like this ?  
case exec_aten::ScalarType::Half:
      logits_f[i] = static_cast<float>(reinterpret_cast<exec_aten::Half*>(logits_last)[i]);
      break;

File: exir/passes/spec_prop_pass.py [hunk 3]
```diff
@@ -9,8 +9,12 @@
 import operator
 from typing import Optional
 
+# register llama.fallback
+import executorch.extension.llm.custom_ops.op_fallback  # noqa: F401
```
Comment (psiddh) [inline exir/passes/spec_prop_pass.py]: Not sure about this import..

a core pass that everyone imports now has a hard dependency on an LLM-specific extension package.

Comment (psiddh) [reply exir/passes/spec_prop_pass.py]: Or maybe like this

  try:
      import executorch.extension.llm.custom_ops.op_fallback  # noqa: F401
  except ImportError:
      pass

**Comment by psiddh** [comment]
lgtm, accepting to unblock

**Review by psiddh — APPROVED** [review]

File: examples/qualcomm/oss_scripts/llama/runner/utils.h (hunk 1 above)
Comment (chenweng-quic) [reply examples/qualcomm/oss_scripts/llama/runner/utils.h:31]: We don't have Half use case now, so not add.

File: examples/qualcomm/oss_scripts/llama/runner/decoder_runner.h (hunk 2 above)
Comment (chenweng-quic) [reply examples/qualcomm/oss_scripts/llama/runner/decoder_runner.h:84]: We don't have Half use case now, so not add.
